rand_core/src/block.rs

```diff
@@ -140,265 +136,95 @@ impl<R: BlockRngCore> BlockRng<R> {
-    ///
-    /// If this is equal to or larger than the size of the result buffer then
-    /// the buffer is "empty" and `generate()` must be called to produce new
-    /// results.
-    #[inline(always)]
-    pub fn index(&self) -> usize {
-        self.index
-    }
 
-    /// Reset the number of available results.
-    /// This will force a new set of results to be generated on next use.
-    #[inline]
-    pub fn reset(&mut self) {
-        self.index = self.results.as_ref().len();
-        self.half_used = false;
+        let res = (self.results.as_ref()[index / 8] >> (index % 8)) & 0b1 != 0;
+        self.index = index + 1;
+        res
     }
 
-    /// Generate a new set of results immediately, setting the index to the
-    /// given value.
-    #[inline]
-    pub fn generate_and_set(&mut self, index: usize) {
-        assert!(index < self.results.as_ref().len());
-        self.core.generate(&mut self.results);
-        self.index = index;
-        self.half_used = false;
-    }
-}
-
-impl<R: BlockRngCore<Item = u64>> RngCore for BlockRng64<R>
-where
-    <R as BlockRngCore>::Results: AsRef<[u64]> + AsMut<[u64]>,
-{
     #[inline]
     fn next_u32(&mut self) -> u32 {
-        let mut index = self.index * 2 - self.half_used as usize;
-        if index >= self.results.as_ref().len() * 2 {
+        let mut index = self.index;
+        index = 4 * ((index / 32) + ((index & 0b1_1111) != 0) as usize);
+
+        if index + 4 >= self.results.as_ref().len() {
             self.core.generate(&mut self.results);
-            self.index = 0;
-            // `self.half_used` is by definition `false`
-            self.half_used = false;
             index = 0;
         }
 
-        self.half_used = !self.half_used;
-        self.index += self.half_used as usize;
-
-        // Index as if this is a u32 slice.
-        unsafe {
-            let results = &*(self.results.as_ref() as *const [u64] as *const [u32]);
-            if cfg!(target_endian = "little") {
-                *results.get_unchecked(index)
-            } else {
-                *results.get_unchecked(index ^ 1)
-            }
-        }
+        let buf = self.results.as_ref()[index..index + 4].try_into().unwrap();
+        self.index = 8 * (index + 4);
+        u32::from_le_bytes(buf)
     }
 
     #[inline]
     fn next_u64(&mut self) -> u64 {
-        if self.index >= self.results.as_ref().len() {
+        let mut index = self.index;
+        index = 8 * ((index / 64) + ((index & 0b11_1111) != 0) as usize);
+
+        if index + 8 >= self.results.as_ref().len() {
             self.core.generate(&mut self.results);
-            self.index = 0;
+            index = 0;
         }
 
-        let value = self.results.as_ref()[self.index];
-        self.index += 1;
-        self.half_used = false;
-        value
+        let buf = self.results.as_ref()[index..index + 8].try_into().unwrap();
+        self.index = 8 * (index + 8);
+        u64::from_le_bytes(buf)
```

Comment:
We always use little-endian order, so this approach will be less performant than the current one on big-endian machines. We could introduce a BE variant of `BlockRng` (or even be generic over endianess), but I doubt it will sufficiently popular.

Reply:
Also see #1026.